### ISSUE TITLE
Allow usage of xDS data from LUA

### ISSUE THREAD

**Issue by bobvanderlinden** [envoyproxy/envoy#14098]
*Title*: Allow usage of xDS data from LUA

*Description*:

Currently we are using LUA to do URL mapping. It requires a large list of URL patterns and for each a replacement pattern. We have around 1000+ of these. It cannot make use of regex, as this would not be performant enough.

The LUA script initially loads a local file (which is filled by `consul-template`) and sets up a lookup tree for the URL patterns. Every time the set of patterns changes, Envoy needs to fully reload in order for LUA to reload the file. Reloading Envoy is not as stable, as triggering reloads too fast would cause `previous envoy process is still initializing`.

Ideally we want LUA to be able to have some kind of xDS service available where we can expose the required data and let LUA reload the lookup-tree when it changes or reload LUA itself with the data being passed to the script. From what I'm gathering this is not possible with the current LUA APIs.

This is a request for allowing LUA to access dynamic data from xDS. However, suggestions are very welcome for an alternative implementation. Should we move to WASM where such features are more viable?

**Comment by dio** [comment]
I think you can send a per-route Lua script via RDS. You can embed the data inside the script, prepared by the xDS Server. Or via LDS if it is a global script.

**Comment by bobvanderlinden** [comment]
This is indeed how we eventually solved it just recently. It does feel a bit dirty, but it seems work quite well.

Our setup now adds lua files in the working directory of Envoy to avoid sending over multiple files. We use ECDS to just update the lua http filter. We eventually chose to use file-based xDS. The file is generated from Consul using consul-template, which writes YAML with a large table in the embedded lua and ends with requiring the local files and passing the table to that module.

What might make the process a bit cleaner is adding initialization parameters alongside the inline script in the configuration. This allows embedding data in yaml/json. This avoids needing to generate safe lua code.

That said, for us this is solved. I think this issue can be closed.
